### PR TITLE
saltstack 2015.2.0rc2 (devel)

### PR DESCRIPTION
**Pull request by DomT4**
* Adds devel block.
* Vendors swig304 due to upstream M2Crypto issues.
* Bump everything to SHA256.
* Modernises the Python arguments.

Closes #38608
Closes https://github.com/Homebrew/homebrew-versions/pull/770

### CODE REVIEW

File: Library/Formula/saltstack.rb [hunk 1]
```diff
@@ -9,80 +10,96 @@ class Saltstack < Formula
...
+    version "2015.2.0rc2"
+  end
 
   depends_on :python if MacOS.version <= :snow_leopard
-  depends_on "swig" => :build
   depends_on "zeromq"
   depends_on "libyaml"
 
+  # For vendored Swig
+  depends_on "pcre"
```
Comment (DomT4) [inline Library/Formula/saltstack.rb]: This may actually only be a ` build ` dependency given ` swig ` is only a build dependency as well.

**Comment by DomT4** [comment]
Failure was OpenSSL. Everything else passed fine.

**Comment by julienlavergne** [comment]
@DomT4  great work for saltstack ! with Boxen it's good for my dev ;)

File: Library/Formula/saltstack.rb [hunk 2]
```diff
@@ -9,80 +10,96 @@ class Saltstack < Formula
...
   # See: https://github.com/martinpaljak/M2Crypto/issues/11
-
   resource "m2crypto" do
     url "https://pypi.python.org/packages/source/M/M2Crypto/M2Crypto-0.22.3.tar.gz"
-    sha1 "c5e39d928aff7a47e6d82624210a7a31b8220a50"
+    sha256 "6071bfc817d94723e9b458a010d565365104f84aa73f7fe11919871f7562ff72"
+  end
+
+  resource "requests" do
+    url "https://pypi.python.org/packages/source/r/requests/requests-2.3.0.tar.gz"
```
Comment (tdsmith) [inline Library/Formula/saltstack.rb]: Does something need a version of requests this old or can we pull a newer one?

Comment (julienlavergne) [inline Library/Formula/saltstack.rb]: You can pull and test a newer one


Julien

**Comment by mikemcquaid** [comment]
@DomT4 thoughts on the audit failure?

**Comment by DomT4** [comment]
It's the old M2Crypto incompatibility with newer OpenSSL versions, meaning we have to tie it to the 098 branch. It can be tied to the 1.0.0.x branch as well, but since that would have to be in Versions we can't do anything in regards to tying this to that.

The OpenSSL compatibility issue has been raised upstream several billion times, but upstream is largely unmaintained these days, it seems, and when it has been talked about they haven't been desperate to fix it.

I *can* make M2Crypto compile with our OpenSSL by doing this:

```ruby
    %w[m2crypto].each do |r|
      resource(r).stage do
        inreplace "setup.py", "self.openssl = '/usr'", "self.openssl = '#{Formula["openssl"].opt_prefix}'"
        system "python", *Language::Python.setup_install_args(libexec/"vendor")
      end
    end
```

But I don't know enough of the Saltstack codebase to check that hasn't screwed something somewhere. I'd need a reproducible test case to work that out.

I think people should find alternatives to M2Crypto to deal with this long-term problem, but alas can't enforce that on upstream.

**Comment by DomT4** [comment]
Going to ping @tdsmith for some Python opinion on whether we can do something with M2Crypto in future:

If I install it with the above block and setup the PYTHONPATH, I can do this fine without any errors:

```
>>> import M2Crypto
>>> from M2Crypto import RSA
>>> from M2Crypto.SSL import Checker
>>> from M2Crypto import X509
>>> from M2Crypto import SSL
```

I can also do:

* ` wget https://raw.githubusercontent.com/M2Crypto/M2Crypto/master/demo/dsa_bench.py `
* ` python -O dsa_bench.py makenewkey sha1 `

And it works:

```
  making and saving a new key
..................+....+++++++++++++++++++++++++++++++++++++++++++++++++++*
...................+........+...+.....+..+.....+.....+....+.....+.....+........................+.+...................+..+.+.....+....+.......+...+...+........+.+....+.........+..+......+.....+...+......+..+....+.........................+..+...+++++++++++++++++++++++++++++++++++++++++++++++++++*
  dsa key length: 1024
  hash algorithm: sha1
  testing signing and verification... ok
  measuring speed...
    100 signings:          0.02s
    100 verifications:     0.02s
```

And:

* ` wget https://raw.githubusercontent.com/M2Crypto/M2Crypto/master/demo/rsa.priv.pem `
* ` wget https://raw.githubusercontent.com/M2Crypto/M2Crypto/master/demo/rsa.pub.pem `
* ` python rsatest.py `:

```
testing public-key encryption: pkcs1_padding
testing public-key encryption: pkcs1_oaep_padding
testing private-key signing: pkcs1_padding
```

All seems to work.

How safe would we be off that success to include the change above in formulae?

**Comment by julienlavergne** [comment]
Any replacement for M2Crypto ?

-- 
Julien Lavergne
Sent with Airmail

On April 19, 2015 at 6:37:15 PM, Dominyk Tiller (notifications@github.com) wrote:

Going to ping @tdsmith for some Python opinion on whether we can do something with M2Crypto in future:

If I install it with the above block and setup the PYTHONPATH, I can do this fine without any errors:

>>> import M2Crypto
>>> from M2Crypto import RSA
>>> from M2Crypto.SSL import Checker
>>> from M2Crypto import X509
>>> from M2Crypto import SSL

I can also do:

wget https://raw.githubusercontent.com/M2Crypto/M2Crypto/master/demo/dsa_bench.py
python -O dsa_bench.py makenewkey sha1
And it works:

  making and saving a new key
..................+....+++++++++++++++++++++++++++++++++++++++++++++++++++*
...................+........+...+.....+..+.....+.....+....+.....+.....+........................+.+...................+..+.+.....+....+.......+...+...+........+.+....+.........+..+......+.....+...+......+..+....+.........................+..+...+++++++++++++++++++++++++++++++++++++++++++++++++++*
  dsa key length: 1024
  hash algorithm: sha1
  testing signing and verification... ok
  measuring speed...
    100 signings:          0.02s
    100 verifications:     0.02s

And:

wget https://raw.githubusercontent.com/M2Crypto/M2Crypto/master/demo/rsa.priv.pem
wget https://raw.githubusercontent.com/M2Crypto/M2Crypto/master/demo/rsa.pub.pem
python rsatest.py:
testing public-key encryption: pkcs1_padding
testing public-key encryption: pkcs1_oaep_padding
testing private-key signing: pkcs1_padding

All seems to work.

How safe would we be off that success to include the change above in formulae?

**Comment by tdsmith** [comment]
Heh, so it's mixing and matching OpenSSL versions somehow if we spec an openssl dependency and don't inreplace the path? If you can make it build, I'm happy.

There's no drop-in replacement for M2Crypto; someone needs to throw money at it or people need to stop using it.

**Comment by DomT4** [comment]
> if we spec an openssl dependency and don't inreplace the path

Yeah, looks like it. M2Crypto seems to expect ` /usr ` unless it's explicitly overwritten, but presumably they are scoping that enough to avoid problems when there's multiple OpenSSLs in the ` $PATH `.

The replacement of ` /usr ` with the desired OpenSSL prefix is also how MacPorts makes it build with newer OpenSSL versions apparently, just checked - They support building Python libraries as ` ports ` which seems like a bad idea to me, but it's a useful reference point to check against.

**Comment by DomT4** [comment]
#38833

**Pull request closed without merging**